src/category_theory/subobject.lean

```diff
@@ -0,0 +1,296 @@
+
+variables {W}
+
+lemma bot_le (X : subobject W) : ⊥ ≤ X :=
+le_of_hom 0 (by simp)
+
+end has_bot
+
+section semilattice_inf
+variables [has_pullbacks C] (W : C)
+
+instance : has_inf (subobject W) :=
+{ inf := λ A B,
+  @mk _ _ _ W (@pullback.fst _ _ _ _ _ A.ι B.ι _ ≫ A.ι) (mono_comp _ _) }
+
+local attribute [instance] mono_comp
+
+lemma le_inf (X Y Z : subobject W) (f : X ≤ Y) (g : X ≤ Z) : X ≤ Y ⊓ Z :=
+le_of_hom (pullback.lift (hom_of_le f) (hom_of_le g) (by simp) ≫ (mk_X_iso _).inv)
+  (by { slice_lhs 2 3 { erw mk_X_iso_inv_comm _, }, simp })
+
+lemma inf_le_left (X Y : subobject W) : X ⊓ Y ≤ X :=
+le_of_hom ((mk_X_iso _).hom ≫ pullback.fst) (by { simp, refl, })
+
+lemma inf_le_right (X Y : subobject W) : X ⊓ Y ≤ Y :=
+le_of_hom ((mk_X_iso _).hom ≫ pullback.snd)
+  (by { rw [category.assoc, ←pullback.condition], simp, refl, })
+
+instance : semilattice_inf (subobject W) :=
+{ le_inf := le_inf W,
+  inf_le_left := inf_le_left W,
+  inf_le_right := inf_le_right W,
+  ..(by apply_instance : partial_order (subobject W)),
+  ..(by apply_instance : has_inf (subobject W)) }
+
+end semilattice_inf
+
+section
+variables [has_coproducts C] [has_images C] (W : C)
+
+instance : has_sup (subobject W) :=
+{ sup := λ A B, mk (image.ι (coprod.desc A.ι B.ι)), }
+
+section
+variables [has_image_maps C]
+
+lemma le_sup_left (X Y : subobject W) : X ≤ X ⊔ Y :=
+le_of_hom
+  ((factor_thru_image coprod.inl ≫ image.map_composable coprod.inl (coprod.desc X.ι Y.ι)) ≫
+    (mk_X_iso (image.ι _)).inv)
+  (by { slice_lhs 3 4 { erw mk_X_iso_inv_comm, }, simp, })
+
+lemma le_sup_right (X Y : subobject W) : Y ≤ X ⊔ Y :=
+le_of_hom
+  ((factor_thru_image coprod.inr ≫ image.map_composable coprod.inr (coprod.desc X.ι Y.ι)) ≫
+    (mk_X_iso (image.ι _)).inv)
+  (by { slice_lhs 3 4 { erw mk_X_iso_inv_comm, }, simp, })
+end
+
+section
+variables [has_equalizers C]
+
+lemma sup_le (X Y Z : subobject W) (f : X ≤ Z) (g : Y ≤ Z) : X ⊔ Y ≤ Z :=
+le_of_hom
+((mk_X_iso _).hom ≫
+  image.eq_to_hom
+    (show coprod.desc X.ι Y.ι = coprod.desc (hom_of_le f) (hom_of_le g) ≫ Z.ι, by simp) ≫
+  image.lift { I := Z.X, m := Z.ι, e := coprod.desc (hom_of_le f) (hom_of_le g) })
+(begin
+  dsimp,
+  rw [category.assoc, category.assoc, image.lift_fac, ←(mk_X_iso _).eq_inv_comp],
+  erw [mk_X_iso_inv_comm, ←image.eq_fac],
+end)
+end
+
+variables [has_image_maps C] [has_equalizers C]
+
+instance : semilattice_sup (subobject W) :=
+{ sup_le := sup_le W,
+  le_sup_left := le_sup_left W,
+  le_sup_right := le_sup_right W,
+  ..(by apply_instance : partial_order (subobject W)),
+  ..(by apply_instance : has_sup (subobject W)) }
+
+end
+
+section
+variables [has_pullbacks C] [has_equalizers C] [has_coproducts C] [has_images C] [has_image_maps C]
+variables (W : C)
+
+instance : lattice (subobject W) :=
+{ ..(by apply_instance : semilattice_inf (subobject W)),
+  ..(by apply_instance : semilattice_sup (subobject W)) }
+
+end
+
+-- PROJECT: Further lattice structures on `subobject W`.
```

Comment:
In the topos project I've shown the complete lattice structure of subobjects in a topos, and the Heyting algebra structure in a locally cartesian closed category already :)

Reply:
Ah, I really should have checked in with you. I didn't intend to do anything more than intersections when I started, but got carried away as more worked with weak assumptions than I'd expected. :-)